src/Microsoft.DotNet.Interactive.Js/tests/client-side-kernel-invocation-context.spec.ts

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { expect } from "chai";
+import { describe } from "mocha";
+import { ClientSideKernelInvocationContext } from "../src/dotnet-interactive/client-side-kernel-invocation-context";
+import { CommandFailedType, CommandSucceededType, DisplayedValueProduced, DisplayedValueProducedType, Disposable, ErrorProduced, ErrorProducedType, KernelCommand, KernelCommandEnvelope, KernelCommandType, KernelEventEnvelope, KernelEventEnvelopeObserver, SubmitCode, SubmitCodeType } from "../src/dotnet-interactive/contracts";
+
+describe("dotnet-interactive", () => {
+    describe("client-side kernel invocation context", () => {
+
+        let makeEventWatcher: () => { watcher: KernelEventEnvelopeObserver, events: KernelEventEnvelope[] } =
+            () => {
+                let events: KernelEventEnvelope[] = [];
+                return {
+                    events,
+                watcher: (ke: KernelEventEnvelope) => events.push(ke)
+                };
+            };
+        function makeSubmitCode(code: string): KernelCommandEnvelope {
+            let command: SubmitCode = {
+                code: code
+            };
+            return {
+                commandType: SubmitCodeType,
+                command: command
+            };
+        }
+        let commandEnvelope: KernelCommandEnvelope = makeSubmitCode("123");
+
+        let toDispose: Disposable[] = [];
+        function use<T extends Disposable>(disposable: T): T {
+            toDispose.push(disposable);
+            return disposable;
+        }
+        afterEach(() => toDispose.forEach(d => d.dispose()));
+
+        it("publishes CommandHandled when Complete is called", async () => {
+            let context = use(ClientSideKernelInvocationContext.establish(commandEnvelope));
```

Comment:
Same observation here, the public api deals in KernelEvents and KernelCommands, the envelopes should come into place near the transport.